phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexToolForDeleteBeforeRebuildIT.java

```diff
@@ -0,0 +1,220 @@
+    }
+
+    @Test
+    /**
+     * Test delete before rebuild
+     */
+    public void testDeleteBeforeRebuildForGlobalIndex() throws Exception {
+        conn.createStatement().execute(String.format(INDEX_GLOBAL_DDL, globalIndexName, dataTableFullName));
+        String globalIndexUpsert = String.format(UPSERT_SQL, globalIndexFullName);
+        PreparedStatement stmt = conn.prepareStatement(globalIndexUpsert);
+        upsertRow(stmt, "tenantID1",11, "name11", 99911);
+        conn.commit();
+
+        ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class).getQueryServices();
+        PTable physicalTable = PhoenixRuntime.getTable(conn, globalIndexFullName);
+        Table hIndexTable= queryServices.getTable(physicalTable.getPhysicalName().getBytes());
+        int count = getUtility().countRows(hIndexTable);
+        // Confirm index has rows.
+        assertEquals(4, count);
+
+        runIndexTool(schemaName, dataTableName, globalIndexName, 0);
+
+        count = getUtility().countRows(hIndexTable);
+
+        // Confirm index has all the data rows
+        assertEquals(3, count);
+    }
+
+    @Test
+    public void testDeleteBeforeRebuildForLocalIndex() throws Exception {
+        String localIndexName = generateUniqueName();
+        conn.createStatement().execute(String.format(INDEX_LOCAL_DDL, localIndexName, dataTableFullName));
+        conn.commit();
+
+        runIndexTool(schemaName, dataTableName, localIndexName, -1);
+    }
+
+    public static String[] getArgValues(String schemaName, String dataTable, String indxTable) {
+        final List<String> args = Lists.newArrayList();
+        if (schemaName != null) {
+            args.add("-s");
+            args.add(schemaName);
+        }
+        args.add("-dt");
+        args.add(dataTable);
+        args.add("-it");
+        args.add(indxTable);
+
+        args.add("-direct");
+        // Need to run this job in foreground for the test to be deterministic
+        args.add("-runfg");
+
+
+        args.add("-deleteall");
+
+        args.add("-op");
+        args.add("/tmp/" + UUID.randomUUID().toString());
+        return args.toArray(new String[0]);
+    }
+
+
+    public static void runIndexTool(String schemaName,
+            String dataTableName, String indexTableName,  int expectedStatus,
+            String... additionalArgs) throws Exception {
+        IndexTool indexingTool = new IndexTool();
+        Configuration conf = new Configuration(getUtility().getConfiguration());
+        conf.set(QueryServices.TRANSACTIONS_ENABLED, Boolean.TRUE.toString());
+        indexingTool.setConf(conf);
+        final String[] cmdArgs =
+                getArgValues(schemaName, dataTableName, indexTableName);
+        List<String> cmdArgList = new ArrayList<>(Arrays.asList(cmdArgs));
+        cmdArgList.addAll(Arrays.asList(additionalArgs));
+        int status = indexingTool.run(cmdArgList.toArray(new String[cmdArgList.size()]));
```

Comment:
Any reason to use the arg list format rather than just passing in parameters?

Reply:
IndexToolIT also has arglist format. Shall I change it to parameter?